pkg-config: fixes for prefixed pkg-config

Mindavi opened the pull request:
###### Description of changes

pkg-config: always prefix with target platform

This reduces the differences between native and cross-compilation.
Due to that it should help preventing some regressions related
to the naming of pkg-config when doing native vs cross-compilation.

Note that some packages may expect pkg-config to not be prefixed,
those will need to be fixed to refer to $PKG_CONFIG rather than
to the hardcoded name pkg-config.

Due to pkg-config not being in the bootstrap packages,
it's relatively easy to change.
For other packages (binutils, gcc) where this change should also
happen, it's a bit harder since the bootstrap binaries are (not yet)
prefixed.

TODO-list:

- [ ] build a lot of packages and see if any of them break
- [ ] write documentation (release notes, usage manual) to make it clear that $PKG_CONFIG should be used and issues may occur when it's hardcoded in a package

Opening as draft since I'm still building some stuff. In principle this seems to work ok, but there'll definitely be packages that are unhappy about this.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @Ericson2314

Mindavi committed:
gtk-doc: use prefixed pkg-config if needed

Needed for the change where we always prefix pkg-config.

Mindavi committed:
ffmpeg: use prefixed pkg-config if needed

Helpful for when we introduce an always prefixed pkg-config.

Mindavi committed:
neon: enable strictDeps

Ericson2314 reviewed (approved):
Neat idea! Unlike the other wrappers, this doesn't also expose the underlying thing, so it won't leak through.

SuperSandro2000 commented:
I don't think this is a good idea like it is right now. People expect pkg-config to be there and needing to patch **everything** will just be a big pain. At least for none cross compilation we should retain pkg-config.

Mindavi commented:
I think this is useful to ensure that any new packages will already think about pkg-config needing to be a prefixed one, leading to less packages or package updates that need patching for cross-compilation. It also makes the difference between 'normal' and 'cross' a bit less, which I think is good to support the usecases better.

I also plan to do this for the (c/c++) compiler (where the benefit of doing this seems quite obvious, an application using `gcc` as compiler won't work on darwin, since it uses `clang`, while forcing the packager to think about setting CC properly during packaging will benefit darwin users).

Also, we cannot _remove_ the prefix when cross-compiling, because we otherwise get a clash in the PATH for the tool names.

But, I think this needs some consideration indeed, and there's definitely packages out there that don't do the right thing with pkg-config. So some fixes here and there are inevitable if we push this through. I'm aware that cross-compilation may be a niche here (and everywhere), but I'd really love if we set good defaults to automatically improve the cross-compilation story rather than having a gap between normal compilation and cross-compilation. I think it'd be really valuable if nixpkgs was the go-to repo for cross-compilation.

If there's opposition to doing this, I'll just merge all the fixes without the change, since the fixes will improve the story for cross-compilation.

Mindavi committed:
cups: support prefixed pkg-config

Otherwise configure fails with the following error:
- configure: error: Need pkg-config to enable libusb support.

SuperSandro2000 commented:
> But, I think this needs some consideration indeed, and there's definitely packages out there that don't do the right thing with pkg-config. So some fixes here and there are inevitable if we push this through.

There will be a bigger amount of packages which just use ``pkg-config`` in some form or another and fixing those won't be a good usage of our time. Sure this would aid cross compiling but the price we pay for it is to high IMO.

Mindavi commented:
On the one hand I do agree that it'll give a lot of headaches for leaf packages, on the other hand it does help ensuring that packages will do more _correct_ packaging.

I'll probably drop the commit though and think if we can do this in another way. I'm not 100% sure how we can enforce this better. Maybe we can enforce it when strictDeps is set? It does kinda make sense to me that we then introduce additional 'strictness' improvements. How does that sound?

SuperSandro2000 commented:
We already have enough trouble with upstream pc and cmake files not using the correct variables. We don't need to add the configure/build system to it.

Mindavi commented:
Removed the pkg-config prefixing and left the patches required to get that to work properly, since they're still needed for proper cross-compilation. Dropped some things that were already fixed in other PRs (mainly neon, I think).

Mindavi commented:
Ok, let's get just those few updates in as a start and leave the unconditional prefixing for another day.

Artturin commented:
The ffmpeg commit broke the cross-compilation of ffmpeg
`pkgsCross.aarch64-multiplatform.ffmpeg
`
```
ERROR: gnutls not found using pkg-config
If you think configure made a mistake, make sure you are using the latest
version from Git.  If the latest version fails, report the problem to the
ffmpeg-user@ffmpeg.org mailing list or IRC #ffmpeg on irc.libera.chat.
Include the log file "ffbuild/config.log" produced by configure as this will help
solve the problem.
```

Mindavi commented:
I know. No idea how I overlooked that. I was planning to revert (at least) that commit but didnt find time yet. Feel free to PR :)

Artturin commented:
https://github.com/NixOS/nixpkgs/pull/200078
